xc/device.ml

```diff
@@ -283,6 +285,37 @@ end
 (****************************************************************************************)
 (** Disks:                                                                              *)
 
+module QMP = struct
+	module Events = struct
+		type event = { domid: int; event: Qmp.event }
+		let qmp_events = Queue.create ()
+		let m, c = Mutex.create(), Condition.create ()
+		let add domid event = Mutex.execute m (fun ()-> Queue.add {domid; event} qmp_events; Condition.signal c)
+		let next () =
+			Mutex.execute m ( fun ()->
+				while Queue.is_empty qmp_events do
+					Condition.wait c m
```

Comment:
This still has the problem in which if we have zero connections to the qmp socket of a domid (so no qmp_read() being executed), we'll wait forever for an event because add() will never be called. Not sure what is the best solution for this. I think that ideally we would need to keep all the fds of all domids open and watch them constantly for an event. Which means we would need to open the connection only once when starting qemu and closing it only once when stopping qemu, and then use poll() instead of Condition.wait() to wait for any of these fds to unblock, indicating a new message is available there to be collected. The problem of this approach is that if we receive a Success/Error message, we won't know which write caused it. I wonder if we could use the ids in each message to match a write to a read.